ykrt/src/mt.rs

```diff
@@ -567,6 +567,34 @@ mod tests {
         );
     }
 
+    #[test]
+    fn dont_trace_two_locations_in_one_thread() {
+        // A thread can only trace one Location at a time: if, having started tracing, it
+        // encounters another Location which has reached its hot threshold, it just ignores it.
+
+        const THRESHOLD: HotThreshold = 5;
+        let mt = MT::new();
+        mt.set_hot_threshold(THRESHOLD);
+        let loc1 = Location::new();
+        let loc2 = Location::new();
+
+        for _ in 0..THRESHOLD {
+            assert_eq!(mt.transition_location(&loc1), TransitionLocation::NoAction);
+            assert_eq!(mt.transition_location(&loc2), TransitionLocation::NoAction);
+        }
+        assert!(matches!(
+            mt.transition_location(&loc1),
+            TransitionLocation::StartTracing(_)
+        ));
+        assert_eq!(mt.transition_location(&loc2), TransitionLocation::NoAction);
+        assert_eq!(
+            hotlocation_discriminant(&loc1),
+            Some(HotLocationDiscriminants::Tracing)
+        );
+        assert!(loc2.load(Ordering::Relaxed).is_counting());
+        assert_eq!(loc2.load(Ordering::Relaxed).count(), THRESHOLD);
```

Comment:
Is it worth checking that `loc2` can be subsequently traced, after loc1 has moved to "compiled"?

Reply:
Not only is that an excellent plan but it highlighted a bug in the transition code. 84f817f minimally fixes it. In essence we had at some point dropped the line of code (in 86b3fbc4d5e8951d9f7cae938168ab079989c193?) that said "this thread is done tracing and can trace another location" so a thread could only trace a single location in its lifetime!

An interesting question is whether:

1. The user should make sure that they don't trace "silly" locations (like the `loop[1|2]` tests do).

or:

2. Before we execute a compiled trace we check to see whether the trace is currently executing something.

Both approaches have pros and cons and, I must admit, I'm not immediately sure which I prefer. Fixing that might be beyond this PR, especially as I have another branch fixing another even nastier bug in the transition code (undefined behaviour) that I think I'd rather focus on first.

Reply:
I'm OK with the fix.

> An interesting question is whether:

I didn't understand this bit, but it sounds like a later PR might explain better?

If so, please squash.

Reply:
> I didn't understand this bit, but it sounds like a later PR might explain better?

I hope so... I need to chew on it!